react-portal Spec (#17824)
microsoft/fluentui: packages/react-portal/Spec.md

#### Focus Management

v0 `Portal` can be configured to focus trap its contents while v8 `Layer` does not offer this and users would need to use `FocusTrapZone` for this purpose.

## Sample Code

`Portal` by default mounts the content to `document.body`. In the event a consumer needs to target a specific mount node for Portal content this should be configurable via prop. Both variants should still be able to access theme and fluent context if available.

```
const customElement = document.createElement('div');

<App> // using FluentProvider of ThemeProvider but not PortalProvider
  <Portal /> // attached to document.body
  <Portal mountNode={customElement} /> // mounted on custom element
</App>
```

`Portal` should be used as a component at any part of the React tree:

```tsx
<ContextProvider>
  <Portal>
    <ContextConsumer /> // should be able to access context
  </Portal>
</ContextProvider>
```

`Portal` should be able to access theme values as css variables:

```tsx
const useStyles = makeStyles({
    portalContent: theme => {...}
})

const styles = useStyles();

<ThemeProvider>

>     <Portal className={styles.portalContent>
>       Can use all theme CSS variables from the parent ThemeProvider
>     </Portal>

Reviewer: (nit) In the current design, the className would have to go on a child element, right?
```suggestion
    <Portal>
      <div className={styles.portalContent}>
        Can use all theme CSS variables from the parent ThemeProvider
      </div>
    </Portal>
```

Author: Yes sorry, there's been a few changes

### v0 migration

- No more openable portals - should use future converged `Popup`
- No more focus trapping in `Portals` do that manually (Tabster)
- No more mount lifecycle methods, users can remedy this easily with `useEffect` or `useLayoutEffect` hooks

## Behaviors

### Server Side Rendering (SSR)

The ReactDOM `createPortal` requires a valid DOM node to render. This is problematic when `document` does not actually exist during the server render. Instead during the server render `null` will be used. This is not a big problem for most components that use portals such as popups or dialogs since they must be opened from some kind of trigger element (i.e. button)

However, there are some cases where a `Portal` content will always need to be rendered on the page. Tooltips should always be rendered so that `aria` attributes will refer to actual elements. This is problematic since the Tooltip (or higher order component) needs to be aware of the server render where `null`is rendered and render the actual content on the first client render.

> The `Portal` component should handle this SSR case, and should be aware of the server and client renders when calling `ReactDOM.createPortal`.

Reviewer: Sorry if I'm missing it, but I don't see details about how Portal will handle SSR? It seems like we need two different behaviors: one that renders the content inline in SSR, and one that doesn't render anything.

One option would be a prop on Portal. But it seems like the more common/desirable case is to render nothing in SSR, so maybe just leave it up to the component to handle inline rendering in SSR. E.g.:
```typescript
{isSSR() ?
  <Tooltip /> /* inline in SSR */ :
  <Portal><Tooltip /></Portal> /* in a portal normally */}
```

Author: So the broad requirement is that `Portal` should be smart enough to handle this IMO, that's the scope of the spec. As a consumer you should be safe knowing that if you use a fluent component, it should be SSR safe. 

On that note I've got an RFC in the pipeline with a concrete solution, stay tuned

### v0/8 Comparison

This section compares the noteworthy differences/similarities between the `Layer` and `Portal` components of v8 and v0 respectively that share functionality with this proposed converged component.

#### Portal Trigger

The v0 Portal component includes a `trigger` prop that allows consumers to open a React portal. The adoption is not widely used across v0's main consumer , Teams.

The v8 `Layer` component fills a similar role to `PortalInner` which is the component that actually renders a React portal.

#### Event propagation

v8 `Layer` users must explicitly use the `enableEventBubbling` behaviour, which is default in React portals. This is mainly due to historical reasons. Meanwhile v0 does not allow any kind of way to disable native React event bubbling through portals, to achieve a similar effect, the user must manage this themselves.

#### DOM insertion

Both `Layer` and `Portal` allow insertion of portals to the same part of the DOM element.

> v7 Layer does this through a `LayerHost` which can be rendered at any part of the React tree. The result is a HTMLDiv element with a specific CSS class an unique Id. `Layer` will attempt to find a `LayerHost` to mount either by CSS class or user provided `hostId`. The default fallback is `document.body`. Each `Layer` will render its own `Fabric` provider.

Reviewer 2: nit:
```suggestion
v8 Layer does this through a `LayerHost` which can be rendered at any part of the React tree. The result is a HTMLDiv element with a specific CSS class and unique Id. `Layer` will attempt to find a `LayerHost` to mount either by CSS class or user provided `hostId`. The default fallback is `document.body`. Each `Layer` will render its own `Fabric` provider.
```

The v0 `PortalBoxContext` stores a single HTMLDiv that is usually a child of `document.body` where all `Portal` components are rendered by default. The v0 `Provider` is rendered for the default portal element, where style and RTL overrides could be applied for all portals within.

#### Lifecycle methods

v0 `Portal` only supports the following lifecycle methods:

- onMount
- onUnmount

`Layer` supports the following:

- onLayerMounted
- onLayerWillUnmount

> The lifecycle events are very similar, with the only difference that `onLayerWillUnmount` being called with `hostId` changes. v0 does not consider the mountNode changing as an unMount of the component itself.

Reviewer 2: nit:
```suggestion
The lifecycle events are very similar, with the only difference being that `onLayerWillUnmount` is called with `hostId` changes. v0 does not consider the mountNode changing as an unmount of the component itself.
```

#### Focus Management

> v0 `Portal` can be configured to focus trap its contents while v7 `Layer` does not offer this and users would need to use `FocusTrapZone` for this purpose.

Reviewer 2: nit:
```suggestion
v0 `Portal` can be configured to focus trap its contents while v8 `Layer` does not offer this and users would need to use `FocusTrapZone` for this purpose.
```

## Sample Code

> `Portal` by default run in a designated area set by a `PortalProvider` or a consumer designated node.

Reviewer 2: nit:
```suggestion
`Portal` by default runs in a designated area set by a `PortalProvider` or a consumer designated node.
```

```tsx
const element = usePortalElement({ targetDocument, className, dir });
const customElement = document.createElement('div');

<PortalContextProvider value={element}>
  <Portal />
  <Portal element={customElement} />
</PortalContextProvider>;
```

> `Portal` should have a fallback of document.body if `PortalProvider` is not used. This fallback should still be able to access theme and fluent context if available

Reviewer 2: nit:
```suggestion
`Portal` should have a fallback of `document.body` if `PortalProvider` is not used. This fallback should still be able to access theme and fluent context if available.
```

```
<App> // using FluentProvider of ThemeProvider but not PortalProvider
  <Portal /> // attached to document.body
  <Portal /> // attached to document.body
</App>
```

> `Portal` should be used as a component at any part of the React tree

Reviewer 2: ```suggestion
`Portal` should be used as a component at any part of the React tree:
```

```tsx
<ContextProvider>
  <Portal>
    <ContextConsumer /> // should be able to access context
  </Portal>
</ContextProvider>
```

> `Portal` should be able to access theme values as css variables

Reviewer 2: ```suggestion
`Portal` should be able to access theme values as css variables:
```

```tsx
const useStyles = makeStyles({
    portalContent: theme => {...}
})

const styles = useStyles();

<ThemeProvider>
    <Portal className={styles.portalContent>
      Can use all theme CSS variables from the parent ThemeProvider
    </Portal>
</ThemeProvider>
```

## Variants

- Mounting the portal on a custom node
- Disable event bubbling

## API

### Portal
> TODO clear whether `onMount` or `onUnmount` need to be called when mount node changes like in v8
>
> TODO clear whether `insertionOrder` should be supported in convergence
>

> > TODO clear whether `disableEventBubbling` should be supported in convergence

Reviewer 2: 👍

| Name                 | Description                                                   | Required | Type              | Default value                    |
| -------------------- | ------------------------------------------------------------- | -------- | ----------------- | -------------------------------- |
| mountNode            | Where the portal is mounted to the DOM                        | No       | HTMLElement       | ProviderContext or document.body |
| onMount              | Called when the portal is mounted                             | No       | Function          |                                  |
| onUnmount            | Called when the portal is unmounted                           | No       | Function          |                                  |

> | disableEventBubbling | Disables event bubbling to the React tree                     | No       | Boolean           |                                  |

Reviewer 2: ```suggestion
| disableEventBubbling | Disables event bubbling to the React tree                     | No       | Boolean           | false                            |
```

| insertionOrder       | Position of the portal content in the mountNode               | No       | 'first' \| 'last' | 'last'                           |

### PortalContext

> Context that store a mount node passed to portals below the provider in the React tree, can be easily accessed with `React.useContext`;

Reviewer 2: nit:
```suggestion
Context that stores the mount node passed to portals below the provider in the React tree, can be easily accessed with `React.useContext`;
```

| Name  | Description                       | Required | Type        |
| ----- | --------------------------------- | -------- | ----------- |
| value | The DOM element to insert portals | Yes      | HTMLElement |

## Structure

Usage with provider

```
<FluentProvider
  <ThemeProvider>
    <PortalProvider>
      <Portal id="portal-1" />
      <Portal id="portal-2" />
    </PortalProvider>
  </ThemeProvider>

> </FluentProvider

Reviewer 2: What's the difference between `FluentProvider` and `ThemeProvider` here?

Author: It's supposed to show that both `FluentProvider` and `ThemeProvider` can apply theme... but you're right and it's confusing... will remove

```suggestion
<FluentProvider
  <PortalProvider>
    <Portal id="portal-1" />
    <Portal id="portal-2" />
  </PortalProvider>
</FluentProvider
```

```

DOM output:
```tsx
<div>Maintree</div>

<div> // PortalProvider
  <div id="portal-1" class="theme-provider-0"}>{children}</div>
  <div id="portal-2" class="theme-provider-0"}>{children}</div>
</div>

```

Usage without provider

```
<FluentProvider
  <ThemeProvider>
    <Portal id="portal-1" />
    <Portal id="portal-2" />
  </ThemeProvider>
</FluentProvider
```

DOM output:
```tsx
<body>

  <div>Maintree</div>

  <div id="portal-1" class="theme-provider-0"}>{children}</div>
  <div id="portal-2" class="theme-provider-0"}>{children}</div>

</body>
```

## Migration

_Describe what will need to be done to upgrade from the existing implementations:_

### v8 migration

> - `enableEventBubbling` will be default, rather use `disabledEventBubbling`

Reviewer 2: ```suggestion
- `enableEventBubbling` will be default, use `disableEventBubbling` instead
```

- No more concept of `LayerHost` and id/class selectors, `PortalContext` should be used to use a raw HTML element
- `insertionOrder` prop will replace `insertFirst`

### v0 migration

- No more openable portals - should use future converged `Popup`
- No more focus trapping in `Portals` do that manually (Tabster)

## Behaviors

No noticeable behaviours, the v0 focus trap functionality should be avoided in favour of manually setting any kind of focus management.

## Accessibility

> This component is considered a utility to render an out of order DOM element. The component will support DOM attribute spreading and `as` composition, therefore any Aria requirements should be handled by the consumer using this component.

Reviewer 2: nit:
```suggestion
This component is considered a utility to render an out of order DOM element. The component will support DOM attribute spreading and `as` composition, therefore any ARIA requirements should be handled by the consumer using this component.
```
